backend/models.py

```diff
@@ -171,11 +178,236 @@ class AttributeDefinition(Base):
+#     user_id = Column(Integer, ForeignKey('users.id'), nullable=False)
+#     permission_id = Column(Integer, ForeignKey('permissions.id'), nullable=False)
+
+#     user = relationship('User')
+#     permission = relationship('Permission')
+
+#     __table_args__ = (
+#         UniqueConstraint('user_id', 'permission_id'),
+#     )
+
+
+# class UserGroup(Base):
+#     __tablename__ = 'user_groups'
+#     id = Column(Integer, primary_key=True)
+#     group_id = Column(Integer, ForeignKey('groups.id'), nullable=False)
+#     user_id = Column(Integer, ForeignKey('users.id'), nullable=False)
+
+#     group = relationship('Group')
+#     user = relationship('User')
+
+#     __table_args__ = (
+#         UniqueConstraint('user_id', 'group_id'),
+#     )
+
+class ChangeStatus(enum.Enum):
+    PENDING = 'PENDING'
+    DECLINED = 'DECLINED'
+    APPROVED = 'APPROVED'
+
+
+class ChangeObject(enum.Enum):
+    SCHEMA = 'SCHEMA'
+    ENTITY = 'ENTITY'
+
+class ChangeType(enum.Enum):
+    CREATE = 'CREATE'
+    UPDATE = 'UPDATE'
+    DELETE = 'DELETE'
+
+class Change(Base):
+    __tablename__ = 'changes'
+    id = Column(Integer, primary_key=True)
+    created_by_user_id = Column(Integer, ForeignKey('users.id'), nullable=False)
+    reviewed_by_user_id = Column(Integer, ForeignKey('users.id'))
+    created_at = Column(DateTime(timezone=True), nullable=False)
+    reviewed_at = Column(DateTime(timezone=True), nullable=True)
+    status = Column(Enum(ChangeStatus), default=ChangeStatus.PENDING, nullable=False)
+    change_object = Column(Enum(ChangeObject), nullable=False)
+    change_type = Column(Enum(ChangeType), nullable=False)
+    comment = Column(String(1024), nullable=True)
+
+    created_by = relationship('User', foreign_keys=[created_by_user_id])
+    reviewed_by = relationship('User', foreign_keys=[reviewed_by_user_id])
+    
+
+class SchemaCreate(Base):
+    __tablename__ = 'schema_create_history'
+    id = Column(Integer, primary_key=True)
+    change_id = Column(Integer, ForeignKey('changes.id'), nullable=False)
+    name = Column(String(128), nullable=False)
+    slug = Column(String(128), nullable=False)
+    reviewable = Column(Boolean, nullable=False)
+
+    change = relationship('Change')
+
+
+class SchemaUpdate(Base):
+    __tablename__ = 'schema_upd_history'
+    id = Column(Integer, primary_key=True) 
+    change_id = Column(Integer, ForeignKey('changes.id'), nullable=False)
+    schema_id = Column(Integer, ForeignKey('schemas.id'), nullable=False)
+    new_name = Column(String(128), nullable=True)
+    old_name = Column(String(128), nullable=True)
+    new_slug = Column(String(128), nullable=True)
+    old_slug = Column(String(128), nullable=True)
+    new_reviewable = Column(Boolean, nullable=True)
+    old_reviewable = Column(Boolean, nullable=True)
+    new_deleted = Column(Boolean, nullable=True)
+    old_deleted = Column(Boolean, nullable=True)
+
+    change = relationship('Change')
+    schema = relationship('Schema')
```

Comment:
Defining the `Change` table and separate tables for every tracked model and action seems robust but a bit difficult to maintain (in case somebody adds/removes columns to/from the models).

Maybe we could use a similar structure inspired by EAV? What would you think of something like this:

![diagram](http://www.plantuml.com/plantuml/svg/XPBDRkem48NtVWgBkyy3iEAgxQABRWhgDfeu9y6jxL1x348LxruhKS4abl0kSyvyyvFP12RFqHeX3gI-g8TSlGqjJ5Z9VrtNoHyXqrE6Gf1RjOUbBF3YMFCv2_ndzhg93779W3jF2jHeayohE-BJdboFTtn711wLwUbsBLyTeuU_nfG7CTgwETSnmDSbZyTHurGmbK34Q_bCZ9ssM13AQc3Y6DPfy6XllC4n7DTyFc0c3Sqd51UxABZLO2rSghjbhrq_nYzYiQv7UnfDVjffiNIdOLf75bcbwT7V_olc7Sd4cLBky2CPaJnPfuDfwp5_YNKuPT93RcgvpCrdoR6pVIRkTCYogkuFK4ptl8nFNx8oS-5wxxSSCLj15JCb7o9Nn0QkZTRy0W00)

Reply:
1. How can this structure be used to track changes in schemas?
2. How can we use it to track creation/deletion of schemas/entities?

It looks like after answering these questions a couple of other tables will be added, leading to very similar model that is there now 🤔 Also, when designing my models I thought that other objects can be tracked (changes in groups, users etc.) and if it so (at least potentially)

Reply:
> Also, when designing my models I thought that other objects can be tracked (changes in groups, users etc.)

Yes, that sounds like a good idea. In the long run we will probably connect to an authentication provider like LDAP or OpenID anyway. I'm not sure how much benefit such a tracking would provide.

However, we actually might be able to track users or other models with my proposal, too. :thinking: 

>  leading to very similar model that is there now

Yes, that could be considered a downside. On the plus side we would not need to worry about conversions between data types; we would also store values as their intended data type.

> How can we use it to track creation/deletion of schemas/entities?

In the `Change` model the type of the change (create, update, delete) would be indicated by the `type` field (or `change_type` if you prefer this name). The model would be specified in the `content_type` field. The `obejct_id` column would store the ID for all actions, except for creations.

Also it would store the column name **or** attribute ID that gets changed. We would also need to store the data type again (not only in `AttributeDefinition`; just in case somebody removes an attribute and adds an attribute with identical name but different data type later).

The `ChangeValueXXX` tables would store the actual old and new values (blank `new_value` for deletions, blank `old_value` for creations).

> How can this structure be used to track changes in schemas?

Indeed, this becomes a little more difficult as schema changes might include changes to attribute definitions. By moving the columns `content_type` and `object_id` to the `Change` model, we could track arbitrary changes to any model.

Let's assume the slug and an attribute definition of a schema change, we would create one `ChangeRequest` instance and

* a `Change` instance, where `content_type` is the Enum for schemas, `object_id` is the schema ID, `fieldname` is "slug" and the related `ChangeValueStr` holds the old/new value.
* a `Change instance`, where `content_type` is the Enum for attribute definitions, `object_id` is the ID of the attribute definition, `fieldname` is the name of the column/field to be changed and the related `ChangeValueXXX` holds the old/new value.

One thing I have not figured out, yet: If an attribute gets removed from a schema, how to keep track of it in the change log of the schema? Maybe it could be as simple as marking attribute definitions as deleted (instead of actually dropping from the table), too? :thinking: 

The updated data model could look like this:

![diagram](http://www.plantuml.com/plantuml/svg/XPF1Si8W48RlF0NZ0-G1FDcfFVJGItLwpMp26cc1M5Xqd8xlNffOkvhjoF7_VpxOXLqaWaZPEwLsq7v0XthsqXjqQQQ_b2whTP2INhLRzC24fq2MZanu88gsoOIIjiID3PPi7zI9dRIEq28qrXNKzNcFMmWTlk9dnaIZConhvlev47OOBreR4GXDtHphd33MKYRYtk912WaWeBbU0k7QUfmIH3SHK4xpqfRi_tYzznXAV4NHXcxGJsfIvRtoXb0SMo7Vkoz2e9gEExovGj-yOqlr_rnfqyQYCm4yyfiF2tuVL2GQ891uUt1vkaFenDjFi6HOHtfdwc7_zBY01sxTuvPIP2wlaMDltgrmnodGLNNcgd4Ac71-qSfLrrzONKQ1XONnOEvDIwhC7LuKSzG2WydUVGC0)

What do you think?

Reply:
Before I answer this question, can you tell me, how to handle attribute renaming in this case?

Reply:
>  how to handle attribute renaming in this case?

Renaming is basically just a replacement of the attribute IDs. To track the change in the schema, the `Change` instance would store the value "attribute_id" in the `fieldname` column. In `ChangeValueInt` we store the old and new ID of the attribute.

I admit, this might look a bit confusion as we also have the `attribute_id` field in the `Change` table :/

Reply:
Do we use `attribute_id` in `Change` only for change of value for user-defined entity fields?

Reply:
That was the idea, yes. But we probably could use `fieldname` for attributes, too :thinking:

Reply:
Okay, so, I guess I wrapped my mind (more or less) around your approach. I guess it can work, probably it's more flexible than current approach🤔 But keep in mind that for it I'll have to rewrite literally all the backend code related to traceability 🤪 I'm okay with that (although it's kind of frustrating), but that will take some time

Reply:
I'm not insisting on changing the approach. Primarily I wanted to share my current impression (not having looked at all the changes in the PR).

If you think it's worth the effort and you would like to change the approach, fine with me. If you prefer to stay on the present course, I'll simply save the idea in an issue in case we ever need to touch this particular part.

Reply:
My reasons to keep current approach (apart from having to rewrite quite an amount of code) is that it's feels...simpler? On the other hand, I wasn't really thinking about someone deleting columns from models (handling adding new columns feels more straightforward), and your approach handles that by not referencing any particular column from any table, which also allows to connect new tables to traceability just by adding new values to some enums. So, if you feel that there will be changes in existing models and/or it will be useful to have such flexibility, that let's (try to) go with this option

Reply:
> that it's feels...simpler? 

Yes, I would have called it "straight forward" and "concrete".

>  So, if you feel that there will be changes in existing models and/or it will be useful to have such flexibility,

I'm not aware of upcoming changes, but who knows what new features, which will require changes to the data model, will be requested in the future :innocent: 

If we do not need to worry about the data model of the tracking tables, it would certainly be helpful. Assuming we ever have to touch one of the tracked tables.

But if you made good progress with the tracking feature, changing direction might not be worth the effort. As I'm not working actively as a developer on the project I can't judge, whether it's worth to change direction now. As primary developer, feel free to make the call. :grin:

Reply:
> whether it's worth to change direction now

Well, currently traceability exists "on its own", meaning that only traceability (for the most part) depends on anything related with it, so these changes will hurt only traceability itself. I don't know if traceability will become "closer" to other components of the application in future, but if this happens and there will be a need to change data model, then it will be a more painful experience than if done now🤔

Reply:
My best guess: the two parts won't get closer by design; most likely future changes might require more model columns being tracked :thinking:

Reply:
Okay, then I guess we should try to go with this approach. I think I won't start reimplementing traceability at least before next week, so maybe you'll have more ideas in the meantime😉

Reply:
Oh, now I'm thinking where to store schema id when creating request for entity creation. Probably it can be placed in `Change.object_id`, but it looks like a misuse of field🤔

Reply:
Sorry for my late reply. But I would also choose the second option :

Reply:
Now I have another issue with creating change request for schema creation. So, among other things, we need to store data of AttributeDefinitions of Schema, which include type, name, flags (required, list, etc.) etc. So, `Change` object for each attr.def. can look like this:
```python
Change(
      change_request=change_request,
      field_name='type',
      value_id=v.id, 
      data_type=ChangeAttrType.STR,
      content_type=ContentType.ATTRIBUTE_DEFINITION,
      change_type=ChangeType.CREATE
)
```
We can easily get `Change` objects for attr.def. creation associated with specific change request, but how to differentiate between different attributes?🤔 The only idea that came to my mind with current data model is to include `object_id=Attribute.id` for each `Change` of each AttributeDefinition, but that would require to create new `Attribute`s during request creation of they don't exist at the moment. What do you think?

Reply:
We had defined that it is very important to never update or delete attributes. The only _risky_ part is that clients create an obscene amount of attributes.

Long story short: For me it would be OK to create `Attribute` instances during request creation.

I could also think of the option to store the attribute name here, but that might complicate things too much :thinking:

Reply:
Ok, so, I am in the middle of the process, but decided to commit what is already there so that you can take a look.
You can look at
* New models for traceability (at the bottom part of `models.py`, line 274)
* All updated functions for entities in `traceability/entity.py`
* `create_schema_create_request` and `apply_schema_create_request` in  `traceability/schema.py`

Reply:
Models look good to me :+1: 

I had a quick look at the functions for entities. I think this is the behavior I had in mind; but I noticed some duplicate or very similar lines, maybe we can reduce code duplication at some time in the future :wink: 

And I noticed that you already avoided this kind of code duplication in `create_schema_create_request` :+1: 

Yes, I think we are on the right track :upside_down_face:

Reply:
Now that's kind of all I wanted to change in traceability (if I didn't forget about anything). A few things I wanted to mention:

* In models we have `ChangeObject` model which was used by initial traceability code extensively. Now it's used just by route for reviewing changes. It can be replaced by `ContentType` model because one is kind of a subset of another, but I'm not sure about it
* In `Change` model we have `value_id` as non-nullable. However there are cases when we don't really need to have value, for example when we are deleting attribute from schema (`traceability/schema.py`, line 337). Maybe it makes sense to make it nullable, but probably the case I described above won't happen too often